apisix/plugins/recaptcha.lua

```diff
@@ -0,0 +1,147 @@
+                        type = "string",
+                        default = "header",
+                        enum = { "header", "query" }
+                    },
+                    param_name = { type = "string", default = "captcha" },
+                }
+            },
+            minItems = 1
+        },
+        response = {
+            type = "object",
+            properties = {
+                content_type = { type = "string", default = "application/json; charset=utf-8" },
+                status_code = { type = "number", default = 400 },
+                body = { type = "string", default = '{"message": "invalid captcha"}' }
+            }
+        },
+
+    },
+    additionalProperties = false,
+    required = { "recaptcha_secret_key" },
+}
+
+local recaptcha_url = "https://www.recaptcha.net"
+
+local _M = {
+    version = 0.1,
+    priority = 700,
+    name = "recaptcha",
+    schema = schema,
+}
+
+function _M.check_schema(conf, schema_type)
+    return core.schema.check(schema, conf)
+end
+
+local function build_radixtree(apis)
+    local items = {}
+    for _, api in ipairs(apis) do
+        local item = {
+            paths = { api.path },
+            methods = api.methods,
+            metadata = api,
+        }
+        table.insert(items, item)
+    end
+    return radix.new(items)
+end
+
+local function find_api(request, apis)
+    local rx = build_radixtree(apis)
+    return rx:match(request.path, { method = request.method })
+end
+
+local function retrieve_captcha(ctx, api)
+    local captcha
+    if api.param_from == "header" then
+        captcha = core.request.header(ctx, api.param_name)
+    elseif api.param_from == "query" then
+        local uri_args = core.request.get_uri_args(ctx) or {}
+        captcha = uri_args[api.param_name]
+    end
+    return captcha
+end
+
+function _M.access(conf, ctx)
+    local path = ctx.var.uri
+    local method = core.request.get_method()
+
+    core.log.debug("path: ", path, ", method: ", method, ", conf: ", core.json.encode(conf))
+
+    local api = find_api({ path = path, method = method }, conf.apis)
+    if not api then
+        return
+    end
+
+    core.log.debug("api found: ", core.json.encode(api))
+
+    local invalid_captcha = true
+    local captcha = retrieve_captcha(ctx, api)
+    if captcha ~= nil and captcha ~= "" then
+        local httpc = http.new()
+        local secret = conf.recaptcha_secret_key
+        local remote_ip = core.request.get_remote_client_ip(ctx)
+        local res, err = httpc:request_uri(recaptcha_url .. "/recaptcha/api/siteverify", {
+            method = "POST",
+            body = "secret=" .. secret .. "&response=" .. captcha .. "&remoteip=" .. remote_ip,
+            headers = {
+                ["Content-Type"] = "application/x-www-form-urlencoded",
+            },
+            ssl_verify = false
+        })
+        if err then
+            core.log.error("request failed: ", err)
+            return 500
```

Comment:
```suggestion
            return 503
```

Reply:
I think here is more like an Internal server error(500) rather than Service Unavailable(503). The recaptcha layer is part of the request, and 500 seems will be more reasonable for the user.

What do you think? 

BTW, Do you think it's a good practice to add a failure tolerance option to let the request continue even recaptcha API is unreachable?

Reply:
> BTW, Do you think it's a good practice to add a failure tolerance option to let the request continue even recaptcha API is unreachable?

IMO, I afraid it's not good practice for this scenario, user should remove the reCAPTCHA plugin instead of tolerate the failure.

Reply:
To make sure we are on the same page, something needs to be clarified. Actually, the behavior of manually removing reCAPTCHA plugin or automatically failure tolerated is just the same. And we are using HTTP to the verify captcha token. Since the network instability, sometimes it fails. the users might won't be able to remove this plugin in time.

Reply:
> IMO, I afraid it's not good practice for this scenario, user should remove the reCAPTCHA plugin instead of tolerate the failure.

To make sure we are on the same page, something needs to be clarified. Actually, the behavior of manually removing reCAPTCHA plugin or automatically failure tolerated is just the same. And we are using HTTP to the verify captcha token. Since the network instability, sometimes it fails. the users might not be able to remove this plugin in time. Just like fault tolerant in the rate-limiting scene.

Reply:
> I think here is more like an Internal server error(500) rather than Service Unavailable(503). The recaptcha layer is part of the request, and 500 seems will be more reasonable for the user.
> 
hi @Dog-Lee When responding with a 500 error, it is intercepted by Nginx's default error_page, so a 503 response would be more appropriate.
Check it out here https://github.com/apache/apisix/blob/c84158c40f1bc3eeb0683310a5ae5e04cce275a9/apisix/cli/ngx_tpl.lua#L315

Reply:
@shuaijinchao Thanks. Then I have no concern about this now.